### ISSUE TITLE
`prefer-type-literal-last` and `undefined` and `null`

### ISSUE THREAD

**Issue by thernstig** [sindresorhus/eslint-plugin-unicorn#3277]
[`prefer-type-literal-last`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v66.0.0/docs/rules/prefer-type-literal-last.md) does not like this:

```ts
  const existingUpload = docs[0] as
    | { id?: number | string | null; url?: string | null }
    | undefined;
```

It wants to change it to:

```ts
  const existingUpload = docs[0] as
    | undefined
    | { id?: number | string | null; url?: string | null };
```

I'd argue having `undefined` or `null` last is much more conventional. An "escape hatch". Should they get special treatment? The rule says:

> Keeping named or otherwise compact types first makes the important top-level type names easier to scan when an inline type literal spans multiple lines.

Why is "compact types first" more important? Isn't this always subjective per code base? I think the rules rationale is not true in my world, as just because I have a type used in one place does not mean I want it first always. At least my gut feel is that, but I am even unsure myself of my thinking here.

_aside: I want to use this issue by saying I am a huge fan of eslint-plugin-unicorn. The rules are awesome. The last release just raised a few eyebrows regarding rule selections, hence why I have been a bit more vocal this last release. I am not complaining in my issues; I am seeking rationale behind rules when I cannot immediately see or understand the rationale given._

**Comment by sindresorhus** [comment]
> Why is "compact types first" more important? Isn't this always subjective per code base? I think the rules rationale is not true in my world, as just because I have a type used in one place does not mean I want it first always. At least my gut feel is that, but I am even unsure myself of my thinking here.

On `null`/`undefined`: youre right, and Ive changed the rule. Placing a nullish "escape hatch" last is the conventional style, so `{ … } | undefined` is now valid and won't be flagged. The enforced order is: named/compact types, then type literals, then `null`/`undefined`.

On "compact types first" more broadly: it isnt about how often a type is used, it's purely about line layout. An inline object type literal usually spans multiple lines, and when it sits first or in the middle of a union it splits the remaining members across the literal's body, so you have to read past the whole object to find out what else the union contains. Keeping the short named members together up front lets you scan them at a glance, with the multi-line literal trailing where it doesn't interrupt the list. A single consistent order also saves re-deciding the layout for every union.

I've also clarified the rationale in the docs.
